### PR TITLE
ci(tracer): fix flaky tracer flare tests

### PR DESCRIPTION
**Pull request by erikayasuda**
## Overview
Some people were seeing flaky behavior with the tracer flare tests (see example [here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/59860/workflows/5b35ea8e-9f63-4e50-972c-0b7b0bbfe6a4/jobs/3765727)).

Seems like these issues were arising from trying to create/destroy/recreate the same `tracer_flare/` directory over and over per test, and sometimes there were race conditions which would fail the assertions to confirm that the directory was cleaned up. 

This PR adds an optional `flare_dir` parameter to the `Flare` init. This is mainly to help with testing, and will default to using the original `tracer_flare` directory name for actual tracer flare jobs. We shouldn't have the same concerns we have with the test race condition for real flare requests, because we will only handle one tracer flare request at a time.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/flare.py [hunk 1]
```diff
@@ -30,22 +30,23 @@
 
 
 class Flare:
-    def __init__(self, timeout_sec: int = DEFAULT_TIMEOUT_SECONDS):
-        self.original_log_level = 0  # NOTSET
-        self.timeout = timeout_sec
+    def __init__(self, timeout_sec: int = DEFAULT_TIMEOUT_SECONDS, flare_dir: str = TRACER_FLARE_DIRECTORY):
+        self.original_log_level: int = logging.NOTSET
+        self.timeout: int = timeout_sec
+        self.flare_dir: pathlib.Path = pathlib.Path(flare_dir)
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:36]: Is this something that can be configured by the user? If not, could we perhaps use a [temporary directory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) instead?

**Review by P403n1x87 — APPROVED** [review]
Thanks for looking into this. I've left just one question, otherwise LGTM. Perhaps as follow-up work it might be good to replace `os.path` with `pathlib` throughout

File: ddtrace/internal/flare.py (hunk 1 above)
Comment (erikayasuda) [reply ddtrace/internal/flare.py:36]: I could maybe use a temp dir just for the tests, but not sure if I can do it for actual tracer flares. I thought about using the temp dir, but I was worried that in the case a customer needs to trigger a tracer flare, they would probably be restarting their app one or more times, in which case the generated files would disappear.

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/internal/flare.py [hunk 2]
```diff
@@ -30,22 +30,23 @@
...
         self.file_handler: Optional[RotatingFileHandler] = None
 
     def prepare(self, configs: List[dict]):
         """
         Update configurations to start sending tracer logs to a file
         to be sent in a flare later.
         """
-        if not os.path.exists(TRACER_FLARE_DIRECTORY):
+        if not os.path.exists(self.flare_dir):
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:44]:
```suggestion
        if not self.flare_dir.exists():
```

File: ddtrace/internal/flare.py [hunk 3]
```diff
@@ -30,22 +30,23 @@
...
             try:
-                os.makedirs(TRACER_FLARE_DIRECTORY)
-                log.info("Tracer logs will now be sent to the %s directory", TRACER_FLARE_DIRECTORY)
+                os.makedirs(self.flare_dir)
+                log.info("Tracer logs will now be sent to the %s directory", self.flare_dir)
             except Exception as e:
-                log.error("Failed to create %s directory: %s", TRACER_FLARE_DIRECTORY, e)
+                log.error("Failed to create %s directory: %s", self.flare_dir, e)
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:49]: Can this block be replaced with
```python
self.flare_dir.mkdir(exists_ok=True)
```
?

File: ddtrace/internal/flare.py [hunk 4]
```diff
@@ -62,7 +63,7 @@ def prepare(self, configs: List[dict]):
 
             ddlogger = get_logger("ddtrace")
             pid = os.getpid()
-            flare_file_path = TRACER_FLARE_DIRECTORY / pathlib.Path(f"tracer_python_{pid}.log")
+            flare_file_path = self.flare_dir / pathlib.Path(f"tracer_python_{pid}.log")
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:66]:
```suggestion
            flare_file_path = self.flare_dir / f"tracer_python_{pid}.log"
```

File: ddtrace/internal/flare.py [hunk 5]
```diff
@@ -133,7 +134,7 @@ def send(self, configs: List[Any]):
                     return
 
     def _generate_config_file(self, pid: int):
-        config_file = TRACER_FLARE_DIRECTORY / pathlib.Path(f"tracer_config_{pid}.json")
+        config_file = self.flare_dir / pathlib.Path(f"tracer_config_{pid}.json")
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:137]:
```suggestion
        config_file = self.flare_dir / f"tracer_config_{pid}.json"
```

File: ddtrace/internal/flare.py [hunk 6]
```diff
@@ -162,8 +163,8 @@ def revert_configs(self):
     def _generate_payload(self, params: Dict[str, str]) -> Tuple[dict, bytes]:
         tar_stream = io.BytesIO()
         with tarfile.open(fileobj=tar_stream, mode="w") as tar:
-            for file_name in os.listdir(TRACER_FLARE_DIRECTORY):
-                flare_file_name = TRACER_FLARE_DIRECTORY / pathlib.Path(file_name)
+            for file_name in os.listdir(self.flare_dir):
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:166]:
```suggestion
            for file_name in self.flare_dir.iterdir():
```

File: ddtrace/internal/flare.py [hunk 7]
```diff
@@ -162,8 +163,8 @@ def revert_configs(self):
...
+                flare_file_name = self.flare_dir / pathlib.Path(file_name)
```
Comment (P403n1x87) [inline ddtrace/internal/flare.py:167]:
```suggestion
            for flare_file_name in self.flare_dir.iterdir():
```

File: tests/internal/test_tracer_flare.py [hunk 8]
```diff
@@ -169,5 +173,5 @@ def test_no_app_logs(self):
         self.flare.revert_configs()
 
     def confirm_cleanup(self):
-        assert not os.path.exists(TRACER_FLARE_DIRECTORY), f"The directory {TRACER_FLARE_DIRECTORY} still exists"
+        assert not os.path.exists(self.flare_dir), f"The directory {self.flare_dir} still exists"
```
Comment (P403n1x87) [inline tests/internal/test_tracer_flare.py:176]:
```suggestion
        assert not self.flare_dir.exists(), f"The directory {self.flare_dir} no longer exists"
```

Comment (erikayasuda) [reply tests/internal/test_tracer_flare.py:176]: I'm taking the proposed change for the assert, but I do want it to flag that the directory still exists if this condition fails

**Pull request merged**